service/storage_service.cc

```diff
@@ -3658,9 +3680,25 @@ future<> storage_service::handle_state_normal(inet_address endpoint, gms::permit
             do_remove_node(endpoint);
         }
     } else if (existing && *existing == endpoint) {
-        tmptr->del_replacing_endpoint(endpoint);
+        // This branch in taken by replace-with-different-ip.
+        // We actually don't remove anything from replacing endpoints here, since
+        // we pass new host_id to del_replacing_endpoint. The same was true
+        // with IP-based token_metadata - the new IP was passed to del_replacing_endpoint
+        // and it didn't remove anything. I'm not sure why we need this code at all.
+        // The old IP is removed by the next block of logic - we detach old IP from token ring,
+        // it gets added to candidates_for_removal, then storage_service::remove_endpoint ->
+        // _gossiper.remove_endpoint -> storage_service::on_remove -> it's removed
+        // from topology and from replacing_endpoints.
+
+        tmptr->del_replacing_endpoint(host_id);
     } else {
-        tmptr->del_replacing_endpoint(endpoint);
+        // This branch is taken by replace-with-same-ip.
+        // We need to grab an old host_id for the given endpoint to properly
+        // remove it from replacing endpoints.
+
+        if (const auto old_host_id = tmptr->get_host_id_if_known(endpoint); old_host_id) {
+            tmptr->del_replacing_endpoint(*old_host_id);
```

Comment:
but... shouldn't the old host_id have been a "being_replaced" endpoint? The **new** host ID is the "replacing endpoint", so this is the one we should call `del_replacing_endpoint` on, no?

Reply:
no, this `del_replacing_endpoint` expects `being_replaced` host_id (or ep in the past) as a parameter:
`void del_replacing_endpoint(host_id existing_node);`